### PR TITLE
[User] Account renewal feature scenarios

### PR DESCRIPTION
**Pull request by TheMadeleine**
| Q                    | A
| -------------       | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | 
| License       | MIT

Based on: #4009

### CODE REVIEW

**Comment by pjedrzejewski** [comment]
@TheMadeleine Please rebase, thank you!

File: features/user/renewing_account_after_deletion.feature [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    As a Visitor
+    I want to be able to register again with the same e-mail
+
+    Background:
+        Given the store operates on a single channel in "France"
+        And there was account of "ted@example.com" with password "pswd"
+        And he had an order
+        But his account was deleted
+
+    @ui @javascript
```
Comment (michalmarcinkowski) [inline features/user/renewing_account_after_deletion.feature]: Do we need `@javascript` for this scenario?

File: src/Sylius/Behat/Context/Setup/OrderContext.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    private $orderFactory;
+
+    /**
+     * @param RepositoryInterface $orderRepository
+     * @param SharedStorageInterface $sharedStorage
+     * @param FactoryInterface $orderFactory
+     */
+    public function __construct(
+        RepositoryInterface $orderRepository,
+        SharedStorageInterface $sharedStorage,
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/OrderContext.php]: Shared storage should be first argument.

File: features/user/renewing_account_after_deletion.feature [hunk 3]
```diff
@@ -0,0 +1,16 @@
+@user
+Feature: Registering account again after it has been deleted
+    In order to set up a new account after I deleted it from the system
+    As a Visitor
+    I want to be able to register again with the same e-mail
+
+    Background:
+        Given the store operates on a single channel in "France"
+        And there was account of "ted@example.com" with password "pswd"
+        And he had an order
```
Comment (michalmarcinkowski) [inline features/user/renewing_account_after_deletion.feature]: This should be removed.

File: src/Sylius/Behat/Context/Setup/OrderContext.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+use Behat\Behat\Context\Context;
+use Sylius\Component\Core\Test\Services\SharedStorageInterface;
+use Sylius\Component\Order\Model\OrderInterface;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Sylius\Component\User\Model\UserInterface;
+/**
+ * @author Magdalena Banasiak <magdalena.banasiak@lakion.com>
+ */
+final class OrderContext implements Context
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/OrderContext.php]: After https://github.com/Sylius/Sylius/pull/4175/files#r53758240 this context will be redundant and should be removed.

File: src/Sylius/Behat/Context/Setup/UserContext.php [hunk 5]
```diff
@@ -172,4 +172,22 @@ public function accountWasDeleted($email)
 
         $this->userRepository->remove($user);
     }
+
+    /**
+     * @Given there was account of :email with password :password
+     */
+    public function thereWasAccountOf($email, $password)
+    {
+        $this->thereIsUserIdentifiedBy($email, $password);
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/UserContext.php]: Not sure if we need a separate method for that step. I would add this annotation to `thereIsUserIdentifiedBy` method.

File: src/Sylius/Behat/spec/Context/Ui/UserContextSpec.php [hunk 6]
```diff
@@ -31,9 +32,10 @@ public function let(
         SharedStorageInterface $sharedStorage,
         UserRepositoryInterface $userRepository,
         CustomerShowPage $customerShowPage,
-        LoginPage $loginPage
+        LoginPage $loginPage,
+        RegisterPage $registerPage
     ) {
-        $this->beConstructedWith($sharedStorage, $userRepository, $customerShowPage, $loginPage);
+        $this->beConstructedWith($sharedStorage, $userRepository, $customerShowPage, $loginPage, $registerPage);
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/spec/Context/Ui/UserContextSpec.php]: I don't see any spec for registering user...

File: src/Sylius/Behat/spec/Page/User/RegisterPageSpec.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+    function it_is_symfony_page()
+    {
+        $this->shouldHaveType(SymfonyPage::class);
+    }
+
+    function it_has_route_name()
+    {
+        $this->getRouteName()->shouldReturn('sylius_user_registration');
+    }
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/spec/Page/User/RegisterPageSpec.php]: Were are the specs for `register` and `wasRegistrationSuccessful` methods? It seems like you don't follow BDD :disappointed:

**Comment by TheMadeleine** [comment]
@michalmarcinkowski corrections applied.

File: src/Sylius/Behat/spec/Context/Ui/UserContextSpec.php [hunk 8]
```diff
@@ -104,4 +106,21 @@ function it_checks_if_account_was_deleted(
 
         $this->accountShouldBeDeleted();
     }
+
+    function it_tries_to_register_new_account(
+        $registerPage
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/spec/Context/Ui/UserContextSpec.php]: Should be in one line. Same as above.

Comment (TheMadeleine) [inline src/Sylius/Behat/spec/Context/Ui/UserContextSpec.php]: :+1:

File: etc/behat/services/contexts.xml [hunk 9]
```diff
@@ -19,6 +19,7 @@
         <parameter key="sylius.behat.context.setup.payment.class">Sylius\Behat\Context\Setup\PaymentContext</parameter>
         <parameter key="sylius.behat.context.setup.security.class">Sylius\Behat\Context\Setup\SecurityContext</parameter>
         <parameter key="sylius.behat.context.setup.user.class">Sylius\Behat\Context\Setup\UserContext</parameter>
+        <parameter key="sylius.behat.context.setup.order.class">Sylius\Behat\Context\Setup\OrderContext</parameter>
```
Comment (lchrusciel) [inline etc/behat/services/contexts.xml]: Why did you add this parameter? It is not used in this file.

File: src/Sylius/Behat/Context/Setup/UserContext.php [hunk 10]
```diff
@@ -75,6 +75,7 @@ public function __construct(
 
     /**
      * @Given there is user :email identified by :password
+     * @Given there was account of :email with password :password
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/UserContext.php]: I'm not sure if we don't have a small conflict here. In #4235 I used transformer to transform `:email` into Customer. I'm pretty curious how this sense will be interpreted after merging this two concepts.

File: src/Sylius/Component/Resource/Model/ToggleableTrait.php [hunk 11]
```diff
@@ -10,6 +10,7 @@
  */
 
 namespace Sylius\Component\Resource\Model;
+use Doctrine\Common\Collections\ArrayCollection;
```
Comment (lchrusciel) [inline src/Sylius/Component/Resource/Model/ToggleableTrait.php]: Missing blank line

File: src/Sylius/Component/Resource/Model/ToggleableTrait.php [hunk 12]
```diff
@@ -22,6 +23,10 @@
     protected $enabled = true;
 
     /**
+     * @var ArrayCollection
+     */
+    protected $coll;
+    /**
```
Comment (lchrusciel) [inline src/Sylius/Component/Resource/Model/ToggleableTrait.php]: Missing blank line

File: src/Sylius/Behat/Context/Setup/UserContext.php (hunk 10 above)
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/UserContext.php]: The step `@Given there is user :email identified by :password` is already on the master, does it work fine?

Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/UserContext.php]: Build is green, so it works fine. Thats funny :tada:

File: src/Sylius/Component/Resource/Model/ToggleableTrait.php (hunk 12 above)
Comment (michalmarcinkowski) [inline src/Sylius/Component/Resource/Model/ToggleableTrait.php]: I would say that these changes should be reverted. That was probably committed by accident.

File: src/Sylius/Behat/spec/Context/Ui/UserContextSpec.php [hunk 13]
```diff
@@ -104,4 +106,19 @@ function it_checks_if_account_was_deleted(
...
+        $registerPage->register('ted@example.com')->shouldBeCalled();
+
+        $this->iTryToRegister('ted@example.com');
+    }
+
+    function it_checks_if_account_is_registered($registerPage)
+    {
+        $registerPage->wasRegistrationSuccessful()->shouldBeCalled();
+
+        $this->iShouldBeRegistered();
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/spec/Context/Ui/UserContextSpec.php]: Where is a spec for case then registration wasn't successful?

File: src/Sylius/Behat/spec/Context/Setup/UserContextSpec.php [hunk 14]
```diff
@@ -142,4 +142,15 @@ function it_deletes_user_account(
 
         $this->accountWasDeleted('ted@test.com');
     }
+
+    function it_deletes_user_account(
+        $sharedStorage,
+        $userRepository,
+        UserInterface $user
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/spec/Context/Setup/UserContextSpec.php]: I suppose the arguments will fit in one line

File: src/Sylius/Component/Resource/Model/ToggleableTrait.php (hunk 12 above)
Comment (lchrusciel) [inline src/Sylius/Component/Resource/Model/ToggleableTrait.php]: :+1: I didn't notice that these changes are useless

File: src/Sylius/Component/Product/Model/Product.php [hunk 15]
```diff
@@ -80,6 +80,7 @@ public function __construct()
         $this->variants = new ArrayCollection();
         $this->options = new ArrayCollection();
         $this->createdAt = new \DateTime();
+        $this->coll = new ArrayCollection();
```
Comment (lchrusciel) [inline src/Sylius/Component/Product/Model/Product.php]: What does it mean _coll_? `Coll is an island in the Inner Hebrides, to the west of the isle of Mull.` at least my dictionary says that.

File: src/Sylius/Behat/spec/Page/User/RegisterPageSpec.php [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+class RegisterPageSpec extends ObjectBehavior
+{
+    function let(Session $session, RouterInterface $router)
+    {
+        $this->beConstructedWith($session, [], $router);
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType('Sylius\Behat\Page\User\RegisterPage');
```
Comment (lchrusciel) [inline src/Sylius/Behat/spec/Page/User/RegisterPageSpec.php]: This class has more logic than is specified here.

File: src/Sylius/Component/Resource/Model/ToggleableTrait.php (hunk 12 above)
Comment (TheMadeleine) [inline src/Sylius/Component/Resource/Model/ToggleableTrait.php]: :+1:

File: src/Sylius/Behat/spec/Page/User/RegisterPageSpec.php (hunk 16 above)
Comment (TheMadeleine) [inline src/Sylius/Behat/spec/Page/User/RegisterPageSpec.php]: Yes, it does. But none of the behaviours of the already existent Pages on master is described by specs. That is why I have left it like that.

File: src/Sylius/Behat/Page/User/RegisterPage.php [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    /**
+     * @throws \Exception
+     */
+    public function wasRegistrationSuccessful()
+    {
+        $flashMessage = $this->getDocument()->find('css', '.alert-success');
+
+        if (null === $flashMessage) {
+            throw new \Exception('Account is not registered.');
+        }
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Page/User/RegisterPage.php]: IMO this method should return `boolean`. `return null !== $flashMessage` and then in the context `expect($this->registerPage->wasRegistrationSuccessful)->toBe(true)` like [here](https://github.com/michalmarcinkowski/Sylius/blob/546beba34235673327463e43e9951a1595cadead/src/Sylius/Behat/Context/Ui/CustomerContext.php#L67-L67)

Comment (TheMadeleine) [inline src/Sylius/Behat/Page/User/RegisterPage.php]: :+1:

File: src/Sylius/Behat/spec/Context/Ui/UserContextSpec.php [hunk 18]
```diff
@@ -104,4 +108,26 @@ function it_checks_if_account_was_deleted(
...
+        $registerPage->wasRegistrationSuccessful()->willReturn(true);
+
+        $this->iShouldBeRegistered();
+    }
+
+    function it_checks_if_account_registration_was_not_successful($registerPage)
+    {
+        $registerPage->wasRegistrationSuccessful()->willReturn(false);
+
+        $this->shouldThrow('PhpSpec\Exception\Example\FailureException')->during('iShouldBeRegistered');
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/spec/Context/Ui/UserContextSpec.php]: `FailureException::class`

File: src/Sylius/Behat/Context/Ui/UserContext.php [hunk 19]
```diff
@@ -45,21 +46,29 @@
...
+     * @var RegisterPage
+     */
+    private $registerPage;
+
+    /**
      * @param SharedStorageInterface $sharedStorage
      * @param UserRepositoryInterface $userRepository
      * @param CustomerShowPage $customerShowPage
      * @param LoginPage $loginPage
+     * @param RegisterPage $registerPage
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/UserContext.php]: Since we try to introduce new approach it should be type hinted with an interface. Also check this in spec of this class

File: src/Sylius/Behat/Page/User/RegisterPage.php [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+        $document->fillField('Last name', 'Smith');
+        $document->fillField('Email', $email);
+        $document->fillField('Password', 'pswd');
+        $document->fillField('Verification', 'pswd');
+
+        $document->pressButton('Register');
+    }
+
+    /**
+     * @return bool
```
Comment (lchrusciel) [inline src/Sylius/Behat/Page/User/RegisterPage.php]: `@throws ElementNotFoundException`

File: src/Sylius/Behat/Page/User/RegisterPage.php [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function getRouteName()
+    {
+        return 'sylius_user_registration';
+    }
+
+    /**
+     * @param string $email
```
Comment (lchrusciel) [inline src/Sylius/Behat/Page/User/RegisterPage.php]: `@throws ElementNotFoundException`

File: src/Sylius/Behat/Page/User/RegisterPage.php [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * @author Magdalena Banasiak <magdalena.banasiak@lakion.com>
+ */
+class RegisterPage extends SymfonyPage
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getRouteName()
```
Comment (lchrusciel) [inline src/Sylius/Behat/Page/User/RegisterPage.php]: Should be protected

File: src/Sylius/Behat/spec/Page/User/RegisterPageSpec.php (hunk 16 above)
Comment (Arminek) [inline src/Sylius/Behat/spec/Page/User/RegisterPageSpec.php]: There is to many inheritance on our page objects :(

Comment (TheMadeleine) [inline src/Sylius/Behat/spec/Page/User/RegisterPageSpec.php]: *too much :D

**Comment by pjedrzejewski** [comment]
Thanks Magda! :+1:
